CSG 2D Lattice Support (#31993)
idaholab/moose: framework/doc/content/source/csg/CSGBase.md

To create any `CSGCell`, use the method `createCell` from `CSGBase` which will return a const reference to the `CSGCell` object that is created (`const CSGCell &`).
At the time of calling `createCell`, a unique cell name, the cell region (`CSGRegion`), and an indicator of the fill must be provided.
The `CSGRegion` is defined by boolean combinations of `CSGSurfaces` as described below.

> Three types of cell fills are currently supported: void, material, universe, and lattice.

Author: ```suggestion
Four types of cell fills are currently supported: void, material, universe, and lattice.
```

!alert-end!

### Lattices

A `CSGLattice` is defined as a patterned arrangement of [`CSGUniverse`](#universes) objects and an "outer" to fill the space around lattice elements.
The `CSGBase` class provides support for two types of 2D lattice types: Cartesian and regular hexagonal.

> It is assumed that these two types are in the $x-y$ plane (having a $+z$ norm).

Reviewer: ```suggestion
It is assumed that these two types are in the $x-y$ plane (having a $+z$ normal).
```

And similarly, the outer fill can be set at the time of initialization or set later with the `setLatticeOuter` method.
Below are two examples of creating a `CSGCartesianLattice` and `CSGHexagonalLattice`, both initialized with the set of `CSGUniverse` objects that define the lattice layout.

!listing CSGBaseTest.C start=create a 2x3 lattice of universes end=dims_map

!listing CSGBaseTest.C start=create a 3-ring hexagonal lattice of universes end=dims_map

Custom lattice types can also be defined to be used with the `CSGBase` class.
Information about how to define custom types of lattices can be found in [source/csg/CSGLattice.md].
If using a custom type, the lattice can be created as a unique pointer and added to the base with `addLattice`.
If there are `CSGUniverse` objects in the lattice at the time of adding it to the base, the `CSGUniverse` objects must already exist in the `CSGBase` instance.
An example of using `addLattice` is shown below (`CSGCartesianLattice` would be replaced with the custom type).

!listing CSGBaseTest.C start=create a lattice as a unique pointer end=check that it exists

!alert! note title=2D vs. 3D Lattices

> The `CSGBase` class supports only the creation of 2D lattices. A "3D" lattice can be created by filling multiple 3D cells with 2D lattices and arranging them in layers to form the desired 3D structure.

Author: I don't think this added wording makes sense. The way that is handled is creating planes at the appropriate z-values and creating a 3D cell in CSG. There isn't "extrusion" in CSG because a 2D "cell" is really just an infinite cell in one direction until planes/surfaces are added to cut it off.

!alert-end!

The `CSGLattice` objects can be accessed or updated with the following methods from `CSGBase`:

> - `setLatticeUniverses`: sets the vector of vectors of `CSGUniverse` objects as the lattice layout.

Author: > maybe it would just work with a single vector passed as a 2d vector

I think this would be the appropriate way to do this.

!media large_media/csg/hex_lat_row.png
       id=fig:hex_lat_row
       caption=Example of a 3-ring hexagonal lattice that has the location indices labeled in the $(i, j)$ form.

Convenience methods are provided for `CSGHexagonalLattice` objects to convert between the required $(i, j)$ indices and a ring-based $(r, p)$ indexing scheme.
In the ring-based scheme, the outermost ring is the 0th ring, and the right-most element in the ring is the 0th position of the ring with indices increasing counter-clockwise around the ring.
For the 3-ring lattice above in [!ref](fig:hex_lat_row), the corresponding $(r, p)$ indices would be as shown in [!ref](fig:hex_lat_ring).

!media large_media/csg/hex_lat_ring.png
       id=fig:hex_lat_ring
       caption=Example of a 3-ring hexagonal lattice that has the location indices labeled in the $(r, p)$ form.

For any lattice, the $(r, p)$ index of a universe in the lattice can be retrieved from the $(i, j)$ index by calling the `getRingIndexFromRowIndex` method.
And similarly, if the $(r, p)$ index form is known, the corresponding $(i, j)$ index can be retrieved using the `getRowIndexFromRingIndex` method.

> It is important to note that while these convenience methods exists to convert between the two indexing schemes, the `CSGUniverse` objects in the lattice can only be accessed using the $(i, j)$ index.

Reviewer: ```suggestion
It is important to note that while these convenience methods exist to convert between the two indexing schemes, the `CSGUniverse` objects in the lattice can only be accessed using the $(i, j)$ index.
```

!alert-end!

### Lattices

A `CSGLattice` is defined as a patterned arrangement of [`CSGUniverse`](#universes) objects and an "outer" to fill the space around lattice elements.
To create any type of lattice, the lattice constructor must be called directly to create a unique pointer.
This pointer then has to be passed to the current `CSGBase` instance with `addLattice` (optionally specifying the lattice type) which will then return a const reference to that generated lattice object (`const & CSGLattice` or `const & LatticeType`).

> At the time that `addLattice` is called to add the lattice to the base instance, any `CSGUniverse` objects associated with the lattice must also be in the `CSGBase` instance already.

Reviewer: ```suggestion
At the time that `addLattice` is called to add the lattice to the base instance, any `CSGUniverse` objects associated with the lattice must also be added (with addUniverse for example) to the `CSGBase` instance already.
```

Author: Skipping this suggestion because `addUniverse` isn't exposed through CSGBase in this way. They have to be created from the base instance with [`createUniverse`](https://github.com/kkiesling/moose/blob/csg_lattices/framework/include/csg/CSGBase.h#L224)

The syntax for creating the pointer and adding it is shown below, where `LatticeType` should be replaced with the specific type of the lattice (e.g., `CSGCartesianLattice` or `CSGHexagonalLattice`).

```cpp
// the unique lattice pointer is made first, creating the lattice object of the specified type
std::unique_ptr<LatticeType> lat_ptr = std::make_unique<:LatticeType>(arguments);

> // and then it is explicitly passed to this CSGBase instance, which holds the memory ownership for the object

Reviewer: ```suggestion
// and then it is explicitly passed to this CSGBase instance, which will keep the memory ownership for the object
```
